bigtable-client-core-parent/bigtable-hbase/src/test/java/com/google/cloud/bigtable/hbase/wrappers/veneer/TestDataClientVeneerApi.java

```diff
@@ -226,8 +253,112 @@ public void testReadRows() throws IOException {
   }
 
   @Test
-  public void testReadRowsCancel() throws IOException {
+  public void testReadPaginatedRows() throws IOException {
+    Query query = Query.create(TABLE_ID).rowKey(ROW_KEY);
+    when(mockDataClient.readRowsCallable(Mockito.<RowResultAdapter>any()))
+        .thenReturn(mockStreamingCallable)
+        .thenReturn(mockStreamingCallable);
+
+    doAnswer(
+            new Answer<Void>() {
+              @Override
+              public Void answer(InvocationOnMock invocation) throws Throwable {
+                ((ResponseObserver) invocation.getArgument(1)).onResponse(Result.EMPTY_RESULT);
+                ((ResponseObserver) invocation.getArgument(1)).onResponse(EXPECTED_RESULT);
+                ((ResponseObserver) invocation.getArgument(1)).onResponse(EXPECTED_RESULT);
+                ((ResponseObserver) invocation.getArgument(1)).onComplete();
+                return null;
+              }
+            })
+        .when(mockStreamingCallable)
+        .call(
+            Mockito.any(com.google.cloud.bigtable.data.v2.models.Query.class),
+            Mockito.any(ResponseObserver.class),
+            Mockito.any(GrpcCallContext.class));
+
+    ResultScanner resultScanner = dataClientWrapper.readRows(query.createPaginator(2), 1000);
+    assertResult(Result.EMPTY_RESULT, resultScanner.next());
+    assertResult(EXPECTED_RESULT, resultScanner.next());
+    assertResult(EXPECTED_RESULT, resultScanner.next());
+
+    doAnswer(
+            new Answer<Void>() {
+              @Override
+              public Void answer(InvocationOnMock invocation) throws Throwable {
+                ((ResponseObserver) invocation.getArgument(1)).onComplete();
+                return null;
+              }
+            })
+        .when(mockStreamingCallable)
+        .call(
+            Mockito.any(com.google.cloud.bigtable.data.v2.models.Query.class),
+            Mockito.any(ResponseObserver.class),
+            Mockito.any(GrpcCallContext.class));
+
+    resultScanner.close();
 
+    ResultScanner noRowsResultScanner = dataClientWrapper.readRows(query.createPaginator(2), 1000);
+    assertNull(noRowsResultScanner.next());
+
+    verify(mockDataClient, times(2)).readRowsCallable(Mockito.<RowResultAdapter>any());
+    verify(mockStreamingCallable, times(2))
+        .call(
+            Mockito.any(Query.class),
+            Mockito.any(ResponseObserver.class),
+            Mockito.any(GrpcCallContext.class));
+  }
+
+  @Test
+  public void testReadRowsLowMemory() throws IOException {
```

Comment:
Similar to above, you want to be testing the rpc between pages. so something like this:

```java
 public void testReadRowsLowMemory() throws IOException {
    Query query = Query.create(TABLE_ID);
    when(mockDataClient.readRowsCallable(any(RowResultAdapter.class)))
        .thenReturn(mockStreamingCallable);

    StreamController mockController = Mockito.mock(StreamController.class);
    doAnswer(invocation -> {
          cancelled.set(true);
          return null;
        })
        .when(mockController)
        .cancel();


    // Generate
    doAnswer(
        (Answer<Void>) invocation -> {
          ResponseObserver<Result> observer = invocation.getArgument(1);
          observer.onStart(mockController);

          for(int i=0; i < 1000 && !cancelled.get(); i++) {
            observer.onResponse(createRow(String.format("row%010d", i)));
            Thread.sleep(10);
          }
          observer.onComplete();
          return null;
        })
        .doAnswer(
        (Answer<Void>) invocation -> {
          ResponseObserver<Result> observer = invocation.getArgument(1);
          observer.onComplete();
          return null;
        })
        .when(mockStreamingCallable)
        .call(any(), any(), any());

    ResultScanner resultScanner = dataClientWrapper.readRows(query.createPaginator(100), 3);
    // Consume the stream
    Lists.newArrayList(resultScanner);

    verify(mockStreamingCallable, times(2))
        .call(
            any(Query.class),
            any(ResponseObserver.class),
            any(GrpcCallContext.class));
    assertTrue(cancelled.get());
  }
```

Reply:
Done